packages/liveblocks-core/src/fsm.ts

```diff
@@ -0,0 +1,493 @@
+    }
+  ): this {
+    if (this.runningState !== RunningState.NOT_STARTED_YET) {
+      throw new Error("Already started");
+    }
+
+    for (const src of this.getMatches(nameOrPattern)) {
+      let map = this.allowedTransitions.get(src);
+      if (map === undefined) {
+        map = new Map();
+        this.allowedTransitions.set(src, map);
+      }
+
+      for (const [type, target_] of Object.entries(mapping)) {
+        const targetSpec = target_ as
+          | Target<TContext, TEvent, TState>
+          | null
+          | undefined;
+        this.knownEventTypes.add(type);
+
+        if (targetSpec !== undefined && targetSpec !== null) {
+          // TODO Disallow overwriting when using a wildcard pattern!
+          const targetFn =
+            typeof targetSpec === "function" ? targetSpec : () => targetSpec;
+          map.set(type, targetFn);
+        }
+      }
+    }
+    return this;
+  }
+
+  /**
+   * Like `.addTransition()`, but takes an (anonymous) transition whenever the
+   * timer fires.
+   *
+   * @param stateOrPattern The state name, or state group pattern name.
+   * @param after          Number of milliseconds after which to take the
+   *                       transition. If in the mean time, another transition
+   *                       is taken, the timer will get cancelled.
+   * @param target     The target state to go to.
+   */
+  public addTimedTransition(
+    stateOrPattern: TState | Wildcard<TState>,
+    after: number | ((context: Readonly<TContext>) => number),
+    target: Target<TContext, TimerEvent, TState>
+  ) {
+    return this.onEnter(stateOrPattern, () => {
+      const ms =
+        typeof after === "function" ? after(this.currentContext) : after;
+      const timeoutID = setTimeout(() => {
+        this.transition({ type: "TIMER" }, target);
+      }, ms);
+
+      return () => {
+        clearTimeout(timeoutID);
+      };
+    });
+  }
+
+  private getTargetFn(
+    eventName: TEvent["type"]
+  ): TargetFn<TContext, TEvent, TState> | undefined {
+    return this.allowedTransitions.get(this.currentState)?.get(eventName);
+  }
+
+  /**
+   * Checks to see if the given event can be handled in the current state.
+   *
+   * XXX Not sure if this method will eventually be needed.
+   */
+  public can(eventName: TEvent["type"]): boolean {
+    return this.getTargetFn(eventName) !== undefined;
+  }
+
+  /**
+   * Exits the current state, and executes any necessary cleanup functions.
+   * Call this before changing the current state to the next state.
+   *
+   * @param levels Defines how many "levels" of nesting will be exited. For
+   * example, if you transition from `foo.bar.qux` to `foo.bar.baz`, then
+   * the level is 1. But if you transition from `foo.bar.qux` to `bla.bla`,
+   * then the level is 3.
+   */
+  private exit(levels: number | null) {
+    for (let i = 0; i < (levels ?? this.cleanupStack.length + 1); i++) {
+      this.cleanupStack.pop()?.();
```

Comment:
I don't this will work when levels is not provided because cleanupStack.length gets evaluated during each iteration while `i` keeps increasing and you will only end up emptying half the array. Usually I do `while(myarray.length) array.pop()` but in your case I think you can just set length before the loop and it will work.

Reply:
Ah yes, good point! I pulled it out of the conditional and before the for loop in ec2df9212d9f05ed2f51dea3df5b12588c6a5b3b. Also noticed that the `+ 1` was a remnant of a previous version, but it wasn't needed. It never broke because `[].pop()` happens to be `undefined`. Thanks! 🙏